### PR TITLE
feat(1559): constrain tx gas price == effective gas price

### PR DESCRIPTION
**Pull request by DreamWuGit**
### Description

constrain tx gas price == effective gas tip  for EIP 1559, new fields need more cells so increase `step_width` as well
### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

<hr>

## How to fill a PR description 

Please give a concise description of your PR.

The target readers could be future developers, reviewers, and auditors. By reading your description, they should easily understand the changes proposed in this pull request.

MUST: Reference the issue to resolve

### Single responsability

Is RECOMMENDED to create single responsibility commits, but not mandatory.

Anyway, you MUST enumerate the changes in a unitary way, e.g.

```
This PR contains:
- Cleanup of xxxx, yyyy
- Changed xxxx to yyyy in order to bla bla
- Added xxxx function to ...
- Refactored ....
```

### Design choices

RECOMMENDED to:
- What types of design choices did you face?
- What decisions you have made?
- Any valuable information that could help reviewers to think critically

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs [hunk 1]
```diff
@@ -90,6 +103,18 @@ impl<F: Field> TxEip1559Gadget<F> {
...
 
+            // calculating min(
+            //     gas_tip_cap
+            //     gas_fee_cap - base_fee_per_gas,
+            // );
+            let diff_gas_base_fee = cb.query_word_rlc();
+            let gas_sub_base_fee = AddWordsGadget::construct(cb, [base_fee.clone(), diff_gas_base_fee.clone()], gas_fee_cap.clone());
+            let lt_gas_tip_gas_fee_sub_base_fee = LtWordGadget::construct(cb, &gas_tip_cap, &diff_gas_base_fee);
+            // let effective_gas_price = priority_fee_per_gas + base_fee_per_gas;
+            let priority_fee_per_gas = cb.query_word_rlc();
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:116]: is this priority_fee_per_gas constrained anywhere?

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:116]: constraint in local changes,  pushed.

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs [hunk 2]
```diff
@@ -144,6 +180,29 @@ impl<F: Field> TxEip1559Gadget<F> {
             offset,
             Some(tx.max_priority_fee_per_gas.to_le_bytes()),
         )?;
+
+        let diff_gas_base_fee = tx.max_fee_per_gas - base_fee;
+        let priority_fee_per_gas = base_fee + tx.max_priority_fee_per_gas.min(diff_gas_base_fee);
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs]: Wrong var name! Misleading! priority_fee_per_gas => effective_gas_price

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs [hunk 3]
```diff
@@ -27,6 +28,14 @@ pub(crate) struct TxEip1559Gadget<F> {
     gas_fee_cap: Word<F>,
     // MaxPriorityFeePerGas
     gas_tip_cap: Word<F>,
+    // condition for min(
+    // gas_tip_cap, gas_fee_cap - base_fee_per_gas)
+    lt_gas_tip_gas_fee_sub_base_fee: LtWordGadget<F>,
+    // gas_fee_cap - base_fee_per_gas
+    gas_sub_base_fee: AddWordsGadget<F, 2, true>,
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: nobody can understand what do you mean by "gas_sub_base_fee"!!! Just change it to gas_fee_cap_minus_ base_fee_per_gas!

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: lt_gas_tip_gas_fee_sub_base_fee!  don't omit word! Nobody knows what "gas_tip" "gas_fee"! You omit what is most important!

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: lt_gas_tip_gas_fee_sub_base_fee is exactly priority_fee_per_gas

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: can be renamed priority_fee_per_gas_check?

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs [hunk 4]
```diff
@@ -90,6 +104,22 @@ impl<F: Field> TxEip1559Gadget<F> {
             let gas_fee_cap_lt_base_fee =
                 LtWordGadget::construct(cb, &gas_fee_cap, &base_fee);
 
+            // calculating min(
+            //     gas_tip_cap
+            //     gas_fee_cap - base_fee_per_gas,
+            // );
+            let diff_gas_base_fee = cb.query_word_rlc();
+            let gas_sub_base_fee = AddWordsGadget::construct(cb, [base_fee.clone(), diff_gas_base_fee.clone()], gas_fee_cap.clone());
+            let lt_gas_tip_gas_fee_sub_base_fee = LtWordGadget::construct(cb, &gas_tip_cap, &diff_gas_base_fee);
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:114]: use MinMaxGadget!

**Comment by lispc** [comment]
Add comment somewhere, explain tx_gas_price is looked up from TxTable in zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: gas_tip notation used in L30, also appears elsewhere before.  of course we can rename it.

**Comment by DreamWuGit** [comment]
> Add comment somewhere, explain tx_gas_price is looked up from TxTable in zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

added in this commit [e1a0d3b](https://github.com/scroll-tech/zkevm-circuits/pull/1152/commits/e1a0d3bcf142705e30eb6dee1b893d4ff0415850)

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs]: updated in [e1a0d3b](https://github.com/scroll-tech/zkevm-circuits/pull/1152/commits/e1a0d3bcf142705e30eb6dee1b893d4ff0415850)

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 3 above)
Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: it is "gas_tip_cap" not "gas_tip". No need to omit one word

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 4 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:114]: MinMaxGadget can not works on Word32 directly. it only can handle max 31 bytes.

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 3 above)
Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: https://github.com/ethereum/go-ethereum/blob/82b0dec7135b281c1b03064d50959dc992c2f94f/core/types/tx_dynamic_fee.go#L31

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: TIP: 小费，gas_tip_【cap】: 我愿意【最多】给的小费。
那 gas_tip 不是听起来更像是 实际给的小费吗？

**Review by lispc — APPROVED** [review]

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs:35]: updated to `lt_gas_tip_cap_vs_gas_fee_cap_sub_base_fee` hh

**Pull request merged**